### DISCUSSION TITLE
merge() method to merge two instances of the same model

### DISCUSSION THREAD

**Discussion by alkasm** [pydantic/pydantic#3416] [Feature Request]
First, my motivating use-case. I have two sources of a model instance: a default that an application provides, and an override which gets parsed from a file (if it exists). As a library writer, to merge these two instances, I will have to implement some recursive merger that works on nested dictionaries so that my users don't have to manually implement merging in their application.

Python's `dict.update()` is probably less useful since it's only the top-level fields (and can be implemented rather simply by sending both model instances to dicts, using update, then constructing a new instance). A different example which is maybe more interesting (but still intuitive) is protobuf's [`MergeFrom()` method](https://googleapis.dev/python/protobuf/latest/google/protobuf/message.html#google.protobuf.message.Message.MergeFrom):

> This method merges the contents of the specified message into the current message. Singular fields that are set in the specified message overwrite the corresponding fields in the current message. Repeated fields are appended. Singular sub-messages and groups are recursively merged.

This would work great for my use-case. 

For inspiration, some related Python merging libraries that I found:
* [`deepmerge`](https://github.com/toumorokoshi/deepmerge) "Deepmerge is a flexible library to handle merging of nested data structures in Python (e.g. lists, dicts)."
* [`jsonmerge`](https://github.com/avian2/jsonmerge) "Merge a series of JSON documents." - has multiple built-in [merge strategies](https://github.com/avian2/jsonmerge#merge-strategies), and you can even provide a json schema to use different strategies per-field, which is pretty cool.
* [`hiyapyco`](https://github.com/zerwes/hiyapyco), "A Hierarchical Yaml Python Config"

**Comment by dd-ssc** [comment]
I have a similar use case here, but TBH, I don't see the need for a `merge(...)` function: I use [`model_dump(...)`](https://docs.pydantic.dev/latest/concepts/serialization/#modelmodel_dump) to turn my pydantic data model instances into dicts and merge those using python built-in functionality (typically the `|` operator). Concise and works like a charm ✅.

**Reply by rlkelly** [reply]
that would just overwrite the fields with the latest model.  merge might mean addition of int fields, and things like that.

**Reply by dd-ssc** [reply]
An _addition of int fields_ is well supported by python built-in functionality (e.g. `|`), but [deep merging](https://pypi.org/project/deepmerge) of the dicts is required for anything non-trivial; see [@mccauleyp's code sample](https://github.com/pydantic/pydantic/discussions/3416#discussioncomment-12267413).

**Reply by alkasm** [reply]
I think they meant literal addition of int values, e.g. `{d: 5} + {d: 6} = {d: 11}`; the point being that merge may have different semantics for different value types.

**Reply by dd-ssc** [reply]
Oh, _that_ kind of addition! OK, didn't think of that - I thought of _adding model fields with integer type_.
Thanks for pointing that out.

Your code sample is not a merge, though; in fact, it's not even valid python - missing quotes for starters, but more importantly, `+` is not supported for dicts:

```
>>> {d: 5} + {d: 6}
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    {d: 5} + {d: 6}
     ^
NameError: name 'd' is not defined. Did you mean: 'id'?
>>> {'d': 5} + {'d': 6}
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    {'d': 5} + {'d': 6}
    ~~~~~~~~~^~~~~~~~~~
TypeError: unsupported operand type(s) for +: 'dict' and 'dict'
```

Also, why would anyone try to add numbers in model instance field value - and expect the result to be the sum ?
I can't seem to think of a use case - that does not involve mixing concerns.

Anyway, recipe for merging model instances: Turn them into dicts using `model_dump(...)` and deepmerge those ✅

**Reply by alkasm** [reply]
> Your code sample is not a merge, though; in fact, it's not even valid python - missing quotes for starters, but more importantly, + is not supported for dicts:

I was presenting it to communicate the idea, not as a literal Python example.

> Also, why would anyone try to add numbers in model instance field value - and expect the result to be the sum ?
> I can't seem to think of a use case - that does not involve mixing concerns.

Here's some valid Python that _does_ behave exactly in this way, to motivate:

```python
>>> from collections import Counter
>>> Counter({"a": 5}) + Counter({"a": 6})
Counter({'a': 11})
```

**Comment by mccauleyp** [comment]
Here's an example that's working well for my use-case:
```py
from typing import TypeVar
from pydantic import BaseModel
from deepmerge import always_merger

T = TypeVar("T", bound=BaseModel)


def merge_pydantic_models(base: T, nxt: T) -> T:
    """Merge two Pydantic model instances.

    The attributes of 'base' and 'nxt' that weren't explicitly set are dumped into dicts
    using '.model_dump(exclude_unset=True)', which are then merged using 'deepmerge',
    and the merged result is turned into a model instance using '.model_validate'.

    For attributes set on both 'base' and 'nxt', the value from 'nxt' will be used in
    the output result.
    """
    base_dict = base.model_dump(exclude_unset=True)
    nxt_dict = nxt.model_dump(exclude_unset=True)
    merged_dict = always_merger.merge(base_dict, nxt_dict)
    return base.model_validate(merged_dict)
```

**Comment by cdjameson** [comment]
I've come across use cases like this many times, but I think it should stay outside the scope of Pydantic. Generally speaking, there is no "one correct way" to merge fields of a model. Lists are probably fine, sure, but most everything else will conflict.
